src/main/java/com/sublinks/sublinksapi/interceptors/RateLimiter.java

```diff
@@ -0,0 +1,129 @@
+
+import com.sublinks.sublinksapi.api.lemmy.v3.site.mappers.LocalSiteRateLimitMapper;
+import com.sublinks.sublinksapi.api.lemmy.v3.site.models.LocalSiteRateLimit;
+import com.sublinks.sublinksapi.instance.models.LocalInstanceContext;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.data.redis.RedisSystemException;
+import org.springframework.data.redis.core.ReactiveRedisTemplate;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.web.server.ResponseStatusException;
+import org.springframework.web.servlet.HandlerInterceptor;
+import reactor.core.Disposable;
+import reactor.core.publisher.Mono;
+
+import java.time.Duration;
+import java.time.Instant;
+
+@Component
+public class RateLimiter implements HandlerInterceptor {
+  private String userName;
+  private final ReactiveRedisTemplate<String, String> redis;
+  private final LocalSiteRateLimitMapper limits;
+  private final LocalInstanceContext localInstanceContext;
+
+  private int maxReqs = 0;
+  private int duration = 0;
+
+  @Autowired
+  public RateLimiter(
+      ReactiveRedisTemplate<String, String> redis,
+      LocalSiteRateLimitMapper limits,
+      LocalInstanceContext localInstanceContext
+  ) {
+    this.redis = redis;
+    this.limits = limits;
+    this.localInstanceContext = localInstanceContext;
+  }
+
+  @Override
+  public boolean preHandle(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Object handler
+  ) throws ResponseStatusException {
+    try {
+      userName = request.getUserPrincipal().getName();
+      // request.getUserPrincipal() will throw if user not authed. We will ignore and just
+      // set remote address as key for rate-limiting.
+    } catch (Exception ignored) {}
+
+
+
+    try {
+      String reqPath = request.getRequestURI();
+      LocalSiteRateLimit rateLimit = limits.convert(localInstanceContext);
+      setRateLimitParameters(rateLimit, reqPath);
+    } catch (Exception e) {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "rate_limiter_error");
+    }
+
+    String key = userName != null ? userName : request.getRemoteAddr();
+
+    try {
+      return getInstanceRateLimiting(key, maxReqs, duration, response);
+    } catch (Exception e) {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "rate_limiter_error");
+    }
+  }
+
+  private void setRateLimitParameters(LocalSiteRateLimit rateLimit, String reqPath)
+      throws Exception {
+    if (rateLimit == null) {
+      throw new Exception("rate limit configuration is null");
+    }
+
+    if (reqPath.startsWith("/api/v3/post")) {
+      this.maxReqs = rateLimit.post_per_second();
+      this.duration = rateLimit.post() * rateLimit.post_per_second();
```

Comment:
Most likely this is wrong but perhaps I should set a global RATE_LIMIT_DURATION variable that can be adjusted on an env level. This was just kinda of filler while testing.